src/dft/backends/mklcpu/backward.cpp

```diff
@@ -30,77 +30,236 @@
 #include "oneapi/mkl/dft/descriptor.hpp"
 #include "oneapi/mkl/dft/detail/mklcpu/onemkl_dft_mklcpu.hpp"
 
+#include "dft/backends/mklcpu/mklcpu_helpers.hpp"
+
+// MKLCPU header
+#include "mkl_dfti.h"
+
 namespace oneapi {
 namespace mkl {
 namespace dft {
 namespace mklcpu {
+namespace detail {
 
-// BUFFER version
+// backward a MKLCPU DFT call to the backend, checking that the commit impl is valid.
+template <dft::detail::precision prec, dft::detail::domain dom>
+inline void check_commit(dft::detail::descriptor<prec, dom> &desc) {
+    auto commit_handle = dft::detail::get_commit(desc);
+    if (commit_handle == nullptr || commit_handle->get_backend() != backend::mklcpu) {
+        throw mkl::invalid_argument("DFT", "computer_backward",
+                                    "DFT descriptor has not been commited for MKLCPU");
+    }
+
+    auto mklcpu_desc = reinterpret_cast<detail::mklcpu_desc_t>(commit_handle->get_handle());
+    MKL_LONG commit_status{ DFTI_UNCOMMITTED };
+    DftiGetValue(mklcpu_desc, DFTI_COMMIT_STATUS, &commit_status);
+    if (commit_status != DFTI_COMMITTED) {
+        throw mkl::invalid_argument("DFT", "compute_backward",
+                                    "MKLCPU DFT descriptor was not successfully committed.");
+    }
+}
+
+// Throw an mkl::invalid_argument if the runtime param in the descriptor does not match
+// the expected value.
+template <dft::detail::config_param Param, dft::detail::config_value Expected, typename DescT>
+inline auto expect_config(DescT &desc, const char *message) {
+    dft::detail::config_value actual{ 0 };
+    desc.get_value(Param, &actual);
+    if (actual != Expected) {
+        throw mkl::invalid_argument("DFT", "compute_backward", message);
+    }
+}
+} // namespace detail
 
 //In-place transform
 template <typename descriptor_type, typename data_type>
 ONEMKL_EXPORT void compute_backward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout) {
-    throw mkl::unimplemented("DFT", "compute_backward", "Not implemented for MKLCPU");
+    detail::expect_config<dft::detail::config_param::PLACEMENT, dft::detail::config_value::INPLACE>(
+        desc, "Unexpected value for placement");
+    auto commit_handle = dft::detail::get_commit(desc);
+    detail::check_commit(desc);
+    auto mklcpu_desc = reinterpret_cast<detail::mklcpu_desc_t>(commit_handle->get_handle());
+
+    sycl::queue &cpu_queue{ commit_handle->get_queue() };
+    sycl::buffer<detail::mklcpu_desc_t, 1> mklcpu_desc_buffer {&mklcpu_desc, sycl::range<1>{1}};
+
+    return cpu_queue.submit([&](sycl::handler& cgh){
+        auto desc_acc = mklcpu_desc_buffer.template get_access<sycl::access::mode::read_write>(cgh);
+        auto inout_acc = inout.template get_access<sycl::access::mode::read_write>(cgh);
+        detail::host_task<detail::kernel_name<detail::mklcpu_desc_t>>(cgh, [=](){
+            DftiComputeBackward(*desc_acc.get_pointer(), inout_acc.get_pointer());
+        });
+    }).wait();
```

Comment:
The wait is not needed, as the queue will take care of the syncronization between multiple compute calls.
> The handle is contained within a buffer in the descriptor class. This avoids the need for buffer destruction.
I agree with this but in the presious stories we had moved the descriptor from the descriptor class to commit class because the handle could be different for other implimentaiton. I say lets stick to that as the overall construction / destruction mostly because of the benifits it gives in regards to serialization

Reply:
This is still problematic. Right now, your code is similar to:

```c++
   sycl::buffer<detail::mklcpu_desc_t, 1> mklcpu_desc_buffer{ &mklcpu_desc, sycl::range<1>{ 1 } };
   ....
   cpu_queue.submit([&](sycl::handler &cgh) {
        auto desc_acc = mklcpu_desc_buffer.template get_access<sycl::access::mode::read>(cgh);
        detail::host_task<class host_kernel_back_inplace>(
            cgh, [=]() { ... });
    });
   return;
```

Upon the destruction of `mklcpu_desc_buffer`, when the function returns, the buffer destructor [will block on](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:buf-sync-rules) all the dependent events in the queue. Consequently, this isn't asynchronous.

Reply:
@hjabird, I understand what you are referring to, I have been trying some alternatives
* pass the raw pointer : this is a simple approach but eventually syncronization becomes a issue in the case if the mklcpu::descriptor is freed before a compute, there is as of now no resolution to this.
* increase the lifespan of the desc_buffer : this would be as you suggested to move the descriptor buffer to the descirptor class, but as the queue/backend specific headers are not available and as there is no mechanism to cast a `buffer<void*>` to a `buffer<DFTI_DESCRIPTOR_HANDEL>` its best if the desc_buffer be defined in the commit_derived_impl class, which would be a vaild solution to the problem as well. this is implimented in 9c560c4ca63c13660c160d5e6aeb5213048b5584
* as the get_handle_buffer (new non-virual method) is required to get access to the buffer the derived class type needs to be known at compile time, for which the commit_impl needs to be forward declared and defined (also a part of 9c560c4ca63c13660c160d5e6aeb5213048b5584), please review this bit, and let me know if this refactoring could be done in a better way.

Reply:
Looks fixed to me! 

Opinion: If you have any enthusiasm for changing things left, I think `auto mklcpu_desc_buffer = detail::get_buffer(commit_handle);` is clearer than `sycl::buffer<detail::mklcpu_desc_t, 1> mklcpu_desc_buffer{ detail::get_buffer(commit_handle) };` , but I'm happy for you to resolve this either way.